Make feebumper class stateless

ryanofsky opened the pull request:
Make feebumper methods static and remove stored state in the class.

Having the results of feebumper calls persist in an object makes process
separation between Qt and wallet awkward, because it means the feebumper object
either has to be serialized back and forth between Qt and wallet processes
between fee bump calls, or that the feebumper object needs to stay alive in the
wallet process with an object reference passed back to Qt. It's simpler just to
have fee bumper calls return their results immediately instead of storing them
in an object with an extended lifetime.

In addition to making feebumper methods static, also:

- Move LOCK calls from Qt code to feebumper
- Move TransactionCanBeBumped implementation from Qt code to feebumper
- Rename CFeeBumper class to FeeBumper (every CFeeBumper reference had to be
  updated in this PR anyway so this doesn't increase the size of the diff)

This change was originally part of https://github.com/bitcoin/bitcoin/pull/10244

ryanofsky committed:
[wallet] Change feebumper from class to functions

Change feebumper from a stateful class into a namespace of stateless
functions.

Having the results of feebumper calls persist in an object makes process
separation between Qt and wallet awkward, because it means the feebumper object
either has to be serialized back and forth between Qt and wallet processes
between fee bump calls, or that the feebumper object needs to stay alive in the
wallet process with an object reference passed back to Qt. It's simpler just to
have fee bumper calls return their results immediately instead of storing them
in an object with an extended lifetime.

In addition to making feebumper stateless, also:

- Move LOCK calls from Qt code to feebumper
- Move TransactionCanBeBumped implementation from Qt code to feebumper

dcousens commented:
Why bother with the class?

ryanofsky commented:
> Why bother with the class?
> Practically a namespace at this point?

I don't see a problem with using a class as a namespace, but happy to change if you think it is in bad taste and have a different suggestion. Note that the files are called feebumper.h/cpp so if you want to replace the class with something else, it might involve file renames. This PR is really more concerned with simplifying Qt code and moving code that doesn't belong there out.

sipa commented:
I agree that if all of a class's methods have become `static` it's better to turn it into a namespace of functions, but don't care strongly. If so, I don't think a file rename is necessary.

ryanofsky commented:
> I agree that if all of a class's methods have become static it's better to turn it into a namespace of functions, but don't care strongly. If so, I don't think a file rename is necessary.

Thanks @sipa, removed class without renaming file.

Added commit 1ea35974529622eb16fe5710bee6719dfc205c31 -> 47c5601d5f6a992d9c46d8300de248374d6eb066 ([pr/ipc-bump.6](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-bump.6) -> [pr/ipc-bump.7](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-bump.7), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-bump.6...pr/ipc-bump.7))

promag commented:
> because it means the feebumper object either has to be serialized back and forth between Qt and wallet processes between fee bump calls

That is a problem or inconvenient?

ryanofsky commented:
>> because it means the feebumper object either has to be serialized back and forth between Qt and wallet processes between fee bump calls

> That is a problem or inconvenient?

These should just be plain function calls because there are no advantages to having a stateful FeeBumper object. Making the feebumper class serializable is dumb because it will add new code that send extra data that isn't always needed.

promag commented on src/wallet/feebumper.cpp:
```diff
@@ -43,70 +44,75 @@ int64_t CalculateMaximumSignedTxSize(const CTransaction &tx, const CWallet *pWal
-    nNewFee(0)
+bool TransactionCanBeBumped(CWallet* pWallet, const uint256& txid)
 {
+    LOCK2(cs_main, pWallet->cs_wallet);
```
It's in the original code, but is it necessary to lock `cs_main`? The same applies for other functions.

ryanofsky replied:
> It's in the original code, but is it necessary to lock cs_main? The same applies for other functions.

Good catch! Fixed in 6a0131608e738f7fcf8ae9efb8ea54bf749ffe85. In the other places cs_main is required for preconditionChecks. Note that in #10973, I replace accesses to `cs_main` global throughout wallet code with `ipc_locked` locals, so it is immediately obvious where locking is and isn't required and mistakes like this will not happen in the future (see `ipc_locked` in https://github.com/ryanofsky/bitcoin/blob/pr/wipc-sep/src/wallet/feebumper.cpp).

promag commented on src/wallet/feebumper.h:
```diff
@@ -23,39 +23,34 @@ enum class BumpFeeResult
+BumpFeeResult CreateBumpFeeTransaction(const CWallet* pWallet,
+    const uint256& txid,
+    const CCoinControl& coin_control,
+    CAmount totalFee,
```
At least in the headers we could fix to snake_case, it doesn't make the patch bigger. WDYT?

ryanofsky replied:
> At least in the headers we could fix to snake_case, it doesn't make the patch bigger. WDYT?

Done in db134b451b7c5871e12bf7ef947d7745446e96f4

promag commented on src/wallet/feebumper.h:
```diff
@@ -23,39 +23,34 @@ enum class BumpFeeResult
-    CAmount nOldFee;
-    CAmount nNewFee;
-};
+/* return whether transaction can be bumped */
```
Use doxygen compatible comments as per developer notes?

ryanofsky replied:
> Use doxygen compatible comments as per developer notes?

Done in db134b451b7c5871e12bf7ef947d7745446e96f4

promag commented on src/wallet/feebumper.h:
```diff
@@ -23,39 +23,34 @@ enum class BumpFeeResult
+bool TransactionCanBeBumped(CWallet* pWallet, const uint256& txid);
+
+/* create bumpfee transaction */
+BumpFeeResult CreateBumpFeeTransaction(const CWallet* pWallet,
```
One line?

ryanofsky replied:
> One line?

I use clang-format for everything, and avoid manual formatting, so that's where this comes from. Would prefer not to change unless someone thinks this is a problem.

ryanofsky reviewed (commented):
Added 2 commits 47c5601d5f6a992d9c46d8300de248374d6eb066 -> db134b451b7c5871e12bf7ef947d7745446e96f4 ([pr/ipc-bump.7](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-bump.7) -> [pr/ipc-bump.8](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-bump.8), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-bump.7...pr/ipc-bump.8))

dcousens reviewed (approved):
light utACK

TheBlueMatt commented on src/wallet/feebumper.cpp:
```diff
@@ -70,7 +70,7 @@ BumpFeeResult PreconditionChecks(const CWallet* pWallet, const CWalletTx& wtx, s
 bool TransactionCanBeBumped(CWallet* pWallet, const uint256& txid)
 {
-    LOCK2(cs_main, pWallet->cs_wallet);
+    LOCK(pWallet->cs_wallet);
```
Can you document the exact locking requirements in GetWalletTx in wallet.h or wallet.cpp just to make sure someone doesnt do some crazy-ass change in the future and blow up the lockorder?

TheBlueMatt commented on src/wallet/feebumper.h:
```diff
@@ -23,34 +21,32 @@ enum class BumpFeeResult
-    std::vector<std::string>& vErrors,
-    CAmount& nOldFee,
-    CAmount& nNewFee,
+    CAmount total_fee,
```
Yes, please do this, but also please keep names consistent between the cpp and h.

TheBlueMatt reviewed (commented):
Hmm, yea, I think I'd prefer leaving the BumpFee:: functions in a namespace. I find BumpFee::* to be cleaner than *BumpFee.

ryanofsky reviewed (commented):
Added 3 commits db134b451b7c5871e12bf7ef947d7745446e96f4 -> 2b2158ae5e21927f332c06ed72239bd2f4392dd9 ([pr/ipc-bump.8](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-bump.8) -> [pr/ipc-bump.9](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-bump.9), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-bump.8...pr/ipc-bump.9))

> Hmm, yea, I think I'd prefer leaving the BumpFee:: functions in a namespace. I find BumpFee::* to be cleaner than *BumpFee.

Added in 2b2158ae5e21927f332c06ed72239bd2f4392dd9

TheBlueMatt commented:
utACK 2b2158ae5e21927f332c06ed72239bd2f4392dd9

jnewbery commented:
I like this, but the commits are structured in a way that makes it more difficult to review than necessary. I think you can remove the first commit that changes all the functions to static (since you're removing the class entirely in a later commit). I also think some of the intermediate commits could perhaps be squashed, although I haven't looked to closely at them.

If you're not interested in maintaining this branch, I can take it.

ryanofsky committed:
[trivial] Rename feebumper variables according to project code style

Future PRs will completely refactor this translation unit and touch all
this code so we rename the variables to follow project stlye guidelines
in this preparation commit.

Don't use m_ prefixes for member variables since we're going to remove
the class entirely in the next commits.

ryanofsky committed:
[refactor] Make feebumper namespace

Future commit will remove the FeeBumper class. This commit simply places
everything into a feebumper namespace, and changes the enum class name
from BumpeFeeResult to feebumper::Result.

ryanofsky commented:
Reset 2b2158ae5e21927f332c06ed72239bd2f4392dd9 -> 510e19c8c9d3c3b33a37dc601835f264df0c6685 ([pr/ipc-bump.9](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-bump.9) -> [pr/ipc-bump.10](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-bump.10))

New branch was written by @jnewbery. Has several changes:

- Simplifies history so related changes are better grouped together.
- Rebased on master
- Reverts some unrelated changes suggested by previous reviewers which could go in separate PRs.
- Other minor changes like declaring functions static instead of using an anonymous namespace.

ryanofsky commented:
Updated 510e19c8c9d3c3b33a37dc601835f264df0c6685 -> 2db9e0a3d123dd7c7f5eb74fdefdb5dba189055b ([pr/ipc-bump.10](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-bump.10) -> [pr/ipc-bump.11](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-bump.11))

Another update from @jnewbery, adding to TransactionCanBeBumped to the namespace.

MarcoFalke commented:
Needs rebase

jnewbery commented:
rebased

ryanofsky reviewed (commented):
utACK aed1d90aca81c20c6e982ad567291f3812d47c8f. Only change since last review was the rebase.

MarcoFalke commented on src/qt/walletmodel.cpp:
```diff
@@ -715,23 +709,15 @@ bool WalletModel::bumpFee(uint256 hash)
-        res = feeBump->commit(wallet);
-    }
-    if(!res) {
+    uint256 txid;
```
nit: `txid_dummy`?

MarcoFalke commented on src/qt/walletmodel.cpp:
```diff
@@ -659,45 +659,39 @@ bool WalletModel::abandonTransaction(uint256 hash) const
+    CAmount old_fee;
+    CAmount new_fee;
+    CMutableTransaction mtx;
+    if (feebumper::CreateTransaction(wallet, hash, coin_control, 0 /* totalFee */, errors, old_fee, new_fee, mtx) != feebumper::Result::OK) {
```
nit: `total_fee`

maflcko reviewed (commented):
utACK aed1d90aca81c20c6e982ad567291f3812d47c8f (Also checked that 7c4f0091957d305679546fde592ffa2de2d186fe produces the same binaries on my arch, since it is not a scripted diff.)
